common/beerocks/bcl/include/bcl/beerocks_event_loop.h

```diff
@@ -0,0 +1,157 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _BEEROCKS_EVENT_LOOP_H_
+#define _BEEROCKS_EVENT_LOOP_H_
+
+#include <functional>
+#include <memory>
+
+// Forward declaration
+class Socket;
+
+namespace beerocks {
+
+/**
+ * @brief The event loop is a programming construct that waits for and dispatches events 
+ * in a program.
+ *
+ * It works by making a request to some "event provider" (which generally blocks the request
+ * until an event has arrived), and then it calls the relevant event handler (i.e.: dispatches
+ * the event).
+ *
+ * The event provider used by the event loop follows the file interface, which can be selected
+ * or 'polled' (the Unix system call, not actual polling). The event loop operates
+ * asynchronously with the event originator.
+ *
+ * When the event loop forms the central control flow construct of a program, as it often does,
+ * it may be termed the main loop or main event loop. This title is appropriate because such an
+ * event loop is at the highest level of control within the program.
+ */
+class EventLoop {
+public:
+    /**
+     * @brief Event handler function definition.
+     *
+     * Parameters to the event handler function are:
+     * - The shared pointer to a socket resource where the event was originated at. 
+     * Event handler can perform, for example, accept, read, write and close operations on the socket.
+     * - The event loop where it was caught. Event handler can install new handlers, remove existing
+     * handlers and even ask for loop termination.
+     * 
+     * @returns True on success or false otherwise
+     */
+    using EventHandler = std::function<bool(std::shared_ptr<Socket>, EventLoop &)>;
+
+    /**
+     * Set of event handler functions, one function to handle each possible event happened.
+     * Handlers are grouped into a struct to facilitate passing them as a single parameter to the install and
+     * remove handlers functions of the event loop.
+     * Event handlers are optional and if not set for a given event, that event will be silently
+     * ignored.
+     */
+    struct EventHandlers {
```

Comment:
It's the other way around. The EventLoop is the main class and Handlers are registered with it.
I don't see a reason to make the EventHandlers a pure virtual class. It'll force you to implement it (with all the handlers, if not all are used) multiple times (instead of just passing lambda functions as handlers)